### PR TITLE
GG-31096 Snapshot encryption fb

### PR DESCRIPTION
**Pull request by alievmirza**

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCacheProcessor.java [hunk 1]
```diff
@@ -3751,6 +3795,20 @@ private void checkReadOnlyState(
                 encGrpCnt++;
         }
 
+        if (!isKeysGenerationRequired) {
+            if (keys != null) {
+                ArrayList<byte[]> cacheKeys = new ArrayList<>();
+
+                for (StoredCacheData ccfg : storedCacheDataList)
+                    if (ccfg.config().isEncryptionEnabled())
```
Comment (sk0x50) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCacheProcessor.java:3803]: Please add brackets.

```
    for (StoredCacheData ccfg : storedCacheDataList) {
        if (...)
            ...
    }
```

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/IgniteCacheDatabaseSharedManager.java [hunk 2]
```diff
@@ -1240,7 +1240,7 @@ public void ensureFreeSpace(DataRegion memPlc) throws IgniteCheckedException {
 
             int emptyDataPagesCnt = freeList.emptyDataPages();
 
-            boolean shouldEvict = allocatedPagesCnt > (memorySize / sysPageSize * plcCfg.getEvictionThreshold()) &&
+                boolean shouldEvict = allocatedPagesCnt > (memorySize / sysPageSize * plcCfg.getEvictionThreshold()) &&
```
Comment (sk0x50) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/IgniteCacheDatabaseSharedManager.java:1243]: unnecessary change :)

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/file/EncryptionUtil.java [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+    private final EncryptionSpi encSpi;
+
+    /** Page size. */
+    private final int pageSize;
+
+    /** Extra bytes added by encryption. */
+    private final int encryptionOverhead;
+
+    /**
+     * Array of zeroes to fulfill tail of decrypted page.
```
Comment (sk0x50) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/file/EncryptionUtil.java:39]: Please use a single-line comments.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/file/EncryptedFileIO.java [hunk 4]
```diff
@@ -258,47 +233,13 @@ private void decrypt(ByteBuffer encrypted, ByteBuffer destBuf) throws IOExceptio
...
-            throw new IOException("Content of encrypted page is broken. [StoredCrc=" + storedCrc +
-                ", calculatedCrc=" + crc + "]");
-        }
-
-        int keyId = encrypted.get() & 0xff;
-
-        encrypted.position(encrypted.position() - (encryptedDataSize() + 4 /* CRC size. */ + 1 /* key identifier. */));
-
-        encrypted.limit(encryptedDataSize());
+        int keyId = encrypted.get(encryptedDataSize() + 4 /* CRC size. */) & 0xff;
```
Comment (sk0x50) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/file/EncryptedFileIO.java:236]:
```
        assert encrypted.remaining() >= pageSize;
        assert encrypted.limit() >= pageSize;
```
These conditions are checked by `EncryptionUtil#decrypt()`, so they can be removed here.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/file/EncryptionUtil.java [hunk 5]
```diff
@@ -0,0 +1,154 @@
+/*
+ * Copyright 2019 GridGain Systems, Inc. and Contributors.
```
Comment (sk0x50) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/file/EncryptionUtil.java:2]: 2020

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/file/EncryptionUtil.java [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+        src.position(srcPos + plainDataSize());
+
+        for (int i = 0; i < encryptionOverhead; i++)
+            assert src.get() == 0 : "Tail of src should be empty [i=" + i + ", pageType=" + pageType + "]";
+
+        src.position(srcPos);
+
+        return true;
+    }
+}
```
Comment (sk0x50) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/file/EncryptionUtil.java:154]: Please add a new line.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/wal/serializer/RecordSerializerFactoryImpl.java [hunk 7]
```diff
@@ -45,22 +46,40 @@
...
     /**
-     * @param cctx Cctx.
+     * @param cctx Cache shared context.
+     * @param readTypeFilter Type filter.
      */
     public RecordSerializerFactoryImpl(
         GridCacheSharedContext cctx,
         @Nullable IgniteBiPredicate<WALRecord.RecordType, WALPointer> readTypeFilter
+    ) {
+        this(cctx, readTypeFilter, null);
```
Comment (sk0x50) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/wal/serializer/RecordSerializerFactoryImpl.java:67]: I would prefer to get a key provider here instead of calling the constructor with `null` value `this(cctx, readTypeFilter, cctx.kernalContext().encryption());`
and add the corresponding assert to the below constructor `assert cacheKeyProvider != null : Encryption key's provider must be initialized.`

Anyway, it is up to you.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/file/EncryptionUtil.java [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+        this.zeroes = new byte[encryptionOverhead];
+    }
+
+    /**
+     * @param srcBuf Source buffer.
+     * @param res Destination buffer.
+     * @param grpKey Encryption key.
+     */
+    public void encrypt(ByteBuffer srcBuf, ByteBuffer res, GroupKey grpKey) {
+        assert srcBuf.remaining() >= pageSize;
```
Comment (sk0x50) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/file/EncryptionUtil.java:61]: Please always use the human-readable message with the assertion here and below.
I think it really helps to analyze log files in case of incidents.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCacheProcessor.java [hunk 9]
```diff
@@ -3472,7 +3472,9 @@ private CacheConfiguration getOrCreateConfigFromTemplate(String cacheName) throw
             (grpKeys, masterKeyDigest) -> {
             assert ccfg == null || !ccfg.isEncryptionEnabled() || !grpKeys.isEmpty();
 
-            DynamicCacheChangeRequest req = prepareCacheChangeRequest(
+                byte[] encCacheKey = prepareEncryptionGroupKey(cacheName, ccfg, F.first(grpKeys));
```
[the remark this answers is not in the document]
Comment (vldpyatkov) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCacheProcessor.java:3475]: I don't see that the code leads to exception without the check.
But maybe add worth: assert ccfg.isEncryptionEnabled() || encCacheKey == null

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCacheProcessor.java [hunk 10]
```diff
@@ -3472,7 +3472,9 @@ private CacheConfiguration getOrCreateConfigFromTemplate(String cacheName) throw
...
+
+                DynamicCacheChangeRequest req = prepareCacheChangeRequest(
```
[the remark this answers is not in the document]
Comment (vldpyatkov) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCacheProcessor.java:3477]: Done.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCacheProcessor.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (vldpyatkov) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCacheProcessor.java:3475]: Assertion does not need here, because we can to be in a case where ccfg is null.
